src/component/contents/exploration/DraftEditorBlockNode.react.js

```diff
@@ -0,0 +1,370 @@
+ * @format
+ * @flow
+ *
+ * This is unstable and not part of the public API and should not be used by
+ * production systems. This file may be update/removed without notice.
+ */
+
+'use strict';
+
+import type {BlockNodeRecord} from 'BlockNodeRecord';
+import type ContentState from 'ContentState';
+import type {DraftBlockRenderMap} from 'DraftBlockRenderMap';
+import type {DraftDecoratorType} from 'DraftDecoratorType';
+import type {DraftInlineStyle} from 'DraftInlineStyle';
+import type SelectionState from 'SelectionState';
+import type {BidiDirection} from 'UnicodeBidiDirection';
+
+const DraftEditorNode = require('DraftEditorNode.react');
+const DraftOffsetKey = require('DraftOffsetKey');
+const EditorState = require('EditorState');
+const Immutable = require('immutable');
+const React = require('React');
+const ReactDOM = require('ReactDOM');
+const Scroll = require('Scroll');
+const Style = require('Style');
+const getElementPosition = require('getElementPosition');
+const getScrollPosition = require('getScrollPosition');
+const getViewportDimensions = require('getViewportDimensions');
+const invariant = require('invariant');
+
+const SCROLL_BUFFER = 10;
+
+const {List} = Immutable;
+
+type BlockRenderFn = (block: BlockNodeRecord) => ?Object;
+type BlockStyleFn = (block: BlockNodeRecord) => string;
+
+type Props = {
+  block: BlockNodeRecord,
+  blockProps?: Object,
+  blockRenderMap: DraftBlockRenderMap,
+  blockRendererFn: BlockRenderFn,
+  blockStyleFn: BlockStyleFn,
+  contentState: ContentState,
+  customStyleFn: (style: DraftInlineStyle, block: BlockNodeRecord) => ?Object,
+  customStyleMap: Object,
+  decorator: ?DraftDecoratorType,
+  direction: BidiDirection,
+  editorKey: string,
+  editorState: EditorState,
+  forceSelection: boolean,
+  selection: SelectionState,
+  startIndent?: boolean,
+  tree: List<any>,
+};
+
+/**
+ * Return whether a block overlaps with either edge of the `SelectionState`.
+ */
+const isBlockOnSelectionEdge = (
+  selection: SelectionState,
+  key: string,
+): boolean => {
+  return selection.getAnchorKey() === key || selection.getFocusKey() === key;
+};
+
+/**
+ * We will use this helper to identify blocks that need to be wrapped but have siblings that
+ * also share the same wrapper element, this way we can do the wrapping once the last sibling
+ * is added.
+ */
+const shouldNotAddWrapperElement = (
+  block: BlockNodeRecord,
+  contentState: ContentState,
+): boolean => {
+  const nextSiblingKey = block.getNextSiblingKey();
+
+  return nextSiblingKey
+    ? contentState.getBlockForKey(nextSiblingKey).getType() === block.getType()
+    : false;
+};
+
+const applyWrapperElementToSiblings = (
+  wrapperTemplate: *,
+  Element: string,
+  nodes: Array<React.Node>,
+): Array<React.Node> => {
+  const wrappedSiblings = [];
+
+  // we check back until we find a sibbling that does not have same wrapper
+  for (const sibling: any of nodes.reverse()) {
+    if (sibling.type !== Element) {
```

Comment:
Why are we tagging this block as `any`?  React.Node could be null or a number, amongst other things, so we should insulate against that?

+nit, but since this is so simple why don't we avoid a break?
```
for (const sibling of nodes.reverse()) {
  if (sibling && sibling.type === Element) {
    wrappedSiblings.push(sibling);
  }
}
// or this could be reduced down to a simple filter:
const wrappedSiblings = nodes.reverse().filter(node ==> node && node.type === Element);
```

Reply:
We want to avoid unnecessary iterations on larger documents